cli/command/stack/kubernetes/list.go

```diff
@@ -44,6 +54,56 @@ func getStacks(kubeCli *KubeCli, opts options.List) ([]*formatter.Stack, error)
 	return formattedStacks, nil
 }
 
+func getStacksWithAllNamespaces(dockerCli command.Cli, kubeCli *KubeCli, opts options.List) ([]*formatter.Stack, error) {
+	stacks, err := getStacks(kubeCli, opts)
+	if err == nil || !apierrs.IsForbidden(err) {
+		return stacks, err
+	}
+	nms, err2 := getUserVisibleNamespaces(dockerCli)
+	if err2 != nil {
+		logrus.Debugf("Failed to query user visible namespaces: %s", err2)
+		// ignore this last error and return the previous one instead
+		return nil, err
+	}
+	opts.AllNamespaces = false
+	for _, namespace := range nms.Items {
+		kubeCli.kubeNamespace = namespace.Name
+		ss, err := getStacks(kubeCli, opts)
+		if err != nil {
+			return nil, err
+		}
+		stacks = append(stacks, ss...)
+	}
+	return stacks, nil
+}
+
+func getUserVisibleNamespaces(dockerCli command.Cli) (*core_v1.NamespaceList, error) {
+	host := dockerCli.Client().DaemonHost()
+	endpoint, err := url.Parse(host)
+	if err != nil {
+		return nil, err
+	}
+	endpoint.Scheme = "https"
+	endpoint.Path = "/kubernetesNamespaces"
+	resp, err := dockerCli.Client().HTTPClient().Get(endpoint.String())
```

Comment:
Wondering if there are other things that should be set on the request (content-type JSON, custom headers that are configured in the client-configuration, etc); https://github.com/docker/cli/blob/60930d309c1544cac67891030aac811f1aff82bf/vendor/github.com/docker/docker/client/request.go#L225-L241

Reply:
I don't know, however this would not be an issue were the `Client` performing the request itself.
https://github.com/moby/moby/pull/37071 has been merged in moby, but maybe it's not to late to bring that up?

Reply:
Well, it's a bit of an odd situation; the `/kubernetesNamespaces` endpoint feels like an add-hoc/temporary solution. Some thoughts on that;

- It's undocumented, and not sure how well-defined the endpoint is
- the Engine remote API may perform validations, e.g.
  - correct request-type, respond different based on "accept" headers (plain-text/JSON)
  - version negotiation (not present in the UCP API) (this could be a problem if we used the `docker/client/request` functionality, as it may be done automatically - need to check)

So, should we export the `docker/client/request` `client.get()`, `client.head()` and so on functions? I don't know.
 
On the "plus" side; when using the `docker/client/request` functionality, we would get some things for free (improved/consistent errors if TLS fails, automatic handling of invalid JSON responses, etc.)

I think in an _ideal_ situation, given that UCP defines its own API (which is the Engine remote-API + Additional endpoints/extensions) there would be an official UCP client that we could use.

Reply:
I do agree with @thaJeztah that it is really an odd situation : we end up hacking the `docker/docker` client around for something that isn't even docker api related.

As discussed a bit offline earlier, a cleaner way (*way cleaner*) would have been to have a kubernetes api extension (same as the current "compose on kubernetes" one) : 
- we wouldn't need to hack `docker/docker` client around, we would just consume the go code from the extension (again, same way as "compose on kubernetes")
- it would be "automatically" versionned (because of k8s way of doing things)
- it wouldn't be a direct dependency on UCP as it could be easily deployed on other cluster in the future if we wanted